course5/homework2.2/newton.cc

```diff
@@ -0,0 +1,48 @@
+#include <iostream>
+#include <cstdlib>
+#include <cmath>
+
+const int MAX_DEPTH = 10000;
+
+struct NewtonMetadata {
+        unsigned int recursionDepth = 0;
+        unsigned long firstStackTop = 0;
+        unsigned long lastStackTop = 0;
+};
+
+float newtonSqrtStep(float input, float oldApproximation, struct NewtonMetadata& metadata)
+{
+        metadata.recursionDepth++;
+        if (metadata.recursionDepth >= MAX_DEPTH)
+        {
+                return oldApproximation;
+        }
+        auto newApproximation = oldApproximation - ((oldApproximation * oldApproximation - input) / (2 * oldApproximation));
+        if (metadata.firstStackTop == 0) {
+                metadata.firstStackTop = reinterpret_cast<unsigned long>(&newApproximation);
```

Comment:
I disagree. For pure, untyped memory address you should use ```void*```; this would represent the intention better.